packages/fxa-payments-server/src/components/fields.tsx

```diff
@@ -0,0 +1,236 @@
+import React, { useContext, useCallback, useRef, useEffect, DetailedHTMLProps, FormHTMLAttributes } from 'react';
+import { ReactStripeElements } from 'react-stripe-elements';
+import classNames from 'classnames';
+import { Validator, FieldType } from '../lib/validator';
+import Tooltip from './Tooltip';
+
+export type FormContextValue = { validator: Validator };
+export const FormContext = React.createContext<FormContextValue | null>(null);
+
+export type OnValidateFunction =
+  (value: any) => ({ value: any, error: any });
+
+type FormProps = {
+  children: React.ReactNode,
+  validator: Validator,
+} & DetailedHTMLProps<FormHTMLAttributes<HTMLFormElement>, HTMLFormElement>;
+
+export const Form = (props: FormProps) => {
+  const {
+    children,
+    validator,
+    ...formProps
+  } = props;
+  return (
+    <form {...formProps}>
+      <FormContext.Provider value={{ validator }}>
+        {children}
+      </FormContext.Provider>
+    </form>
+  );
+};
+
+type FieldGroupProps = { children: React.ReactNode };
+
+export const FieldGroup = ({ children }: FieldGroupProps) => (
+  <div className="input-row-group">
+    {children}
+  </div>
+);
+
+type FieldProps = {
+  name: string,
+  initialValue?: any,
+  tooltip?: boolean,
+  required?: boolean,
+  label?: string | React.ReactNode,
+  className?: string,
+};
+
+type FieldHOCProps = {
+  fieldType: FieldType,
+  tooltipParentRef?: React.MutableRefObject<any>,
+  children: React.ReactNode,
+} & FieldProps;
+
+export const Field = ({
+  tooltipParentRef,
+  fieldType,
+  name,
+  initialValue = null,
+  tooltip = true,
+  required = false,
+  label,
+  className='input-row',
+  children,
+}: FieldHOCProps) => {
+  const { validator } = useContext(FormContext) as FormContextValue;
+
+  // Disabling lint rule because registering a field changes the validator
+  // instance, so including validator in list of deps causes an infinite loop
+  /* eslint-disable react-hooks/exhaustive-deps */
+  useEffect(
+    () => validator.registerField({ name, initialValue, required, fieldType }),
+    [ name, required, fieldType ]
+  );
+  /* eslint-enable react-hooks/exhaustive-dep */
+
+  return (
+    <div className={className}>
+      <label>
+        {label && <span className="label-text">{label}</span>}
+        {children}
+        {tooltip && tooltipParentRef && validator.getError(name) &&
+          <Tooltip parentRef={tooltipParentRef}>{validator.getError(name)}</Tooltip>}
+      </label>
+    </div>
+  );
+};
+
+type InputProps =
+  & { onValidate?: OnValidateFunction }
```

Comment:
It's an intersection type, so all the properties of all these types are required. I'm not familiar with the & in front though, maybe that makes it a type literal?

Reply:
I saw this as a formatting pattern elsewhere kind of like trailing commas - turns out you can put `&` in front of all the lines of an intersection type like this, so that you can add & sort lines without worrying which ones had ampersands and not